tpx3awkward/_utils.py

```diff
@@ -93,155 +101,160 @@ def _shift_xy(chip, row, col):
-            # The high 6 bits of the row
-            # '0000000111111000'
-            spix = (l_pix_addr & np.uint(0x01F8)) >> np.uint(1)
-            rowcol = _shift_xy(
-                chip,
-                # add the low 2 bits of the row
-                # '0000000000000011'
-                spix + (l_pix_addr & np.uint(0x3)),
-                # add the low 1 bit of the column
-                # '0000000000000100'
-                dcol + ((l_pix_addr & np.uint(0x4)) >> np.uint(2)),
-            )
-            col = x[photon_offset] = rowcol[1]
-            y[photon_offset] = rowcol[0]
-            # ToA is 14 bits
-            ToA[photon_offset] = (msg >> np.uint(30)) & np.uint(0x3FFF)
-            # ToT is 10 bits
-            # report in ns
-            ToT[photon_offset] = ((msg >> np.uint(20)) & np.uint(0x3FF)) * 25
-            # FToA is 4 bits
-            l_FToA = FToA[photon_offset] = (msg >> np.uint(16)) & np.uint(0xF)
-            # SPIDR time is 16 bits
-            SPIDR[photon_offset] = msg & np.uint(0xFFFF)
-            # chip number (this is a constant)
-            chip_number[photon_offset] = chip
-            # heartbeat time
-            basetime[photon_offset] = heartbeat_time
-
-            ToA_coarse = (SPIDR[photon_offset] << np.uint(14)) | ToA[photon_offset]
-            pixelbits = int((ToA_coarse >> np.uint(28)) & np.uint(0x3))
-            heartbeat_time_bits = int((heartbeat_time >> np.uint(28)) & np.uint(0x3))
-            diff = heartbeat_time_bits - pixelbits
-            if diff == 1 or diff == -3:
-                heartbeat_time -= np.uint(0x10000000)
-            elif diff == -1 or diff == 3:
-                heartbeat_time += np.uint(0x10000000)
-            globaltime = (heartbeat_time & np.uint(0xFFFFC0000000)) | (ToA_coarse & np.uint(0x3FFFFFFF))
-
-            timestamp[photon_offset] = (globaltime << np.uint(12)) - (l_FToA << np.uint(8))
-            # correct for phase shift
-            phase = np.uint((col / 2) % 16)
-            if phase == 0:
-                timestamp[photon_offset] += 16 << 8
-            else:
-                timestamp[photon_offset] += phase << 8
-
-            photon_offset += 1
-            msg_run_count += 1
-        elif typ == 3:
-            #  3: TDC timstamp (id'd via 0x6 upper nibble)
-            # TODO: handle these!
-            msg_run_count += 1
-        elif typ == 4:
-            #  4: global timestap (id'd via 0x4 upper nibble)
-            subheader = (msg >> np.uint(56)) & np.uint(0x0F)
-            if subheader == 0x4:
-                # timer lsb, 32 bits of time
-                heartbeat_lsb = (msg >> np.uint(16)) & np.uint(0xFFFFFFFF)
-            elif subheader == 0x5:
-                # timer msb
-
-                time_msg = (msg >> np.uint(16)) & np.uint(0xFFFF)
-                heartbeat_msb = time_msg << np.uint(32)
-                # TODO the c++ code has large jump detection, do not understand why
```

Comment:
I suspect that this line is the actual bug that needed to be fixed.

Reply:
I'd like to see how the heartbeats are being stored in the actual data (so far I haven't seen any heartbeat messages). @JGoodrichBNL is this the right Run ID to check: 143210? Do you remember by any chance what heartbeat rate was set up in the detector when you acquired those data?

Reply:
Hi @genematx,

No, that scan is quite old.

Here are some scans with heartbeats. Note that the data is very sparse as we did this without beam and the only hits are background/cosmic rays. We will try with beam on Monday or Tuesday during data security time.

All scans below are 100 partitions at at 1.000 sec exposure time/partition with an 1.002 sec acquire time/partition (e.g., 1.000 sec of collection followed by 2 ms deadtime before starting next partition).

152384 - global_timestamp_interval = 500
152389 - global_timestamp_interval = 250
152389 - global_timestamp_interval = 0 (i.e., what all of our old data had)
152396 - global_timestamp_interval = 1
152397 - global_timestamp_interval = 0.1
152401 - global_timestamp_interval = 0.01

We believe the values are in ms, although the timestamp rate in the files does not seem to match this more. @josephhanrahan can elaborate. Perhaps you can upload some of your viewing codes on your branch, Joe?

Please note we are coming up on our beamtime very quickly, so we would like to get this resolved ASAP.

Thanks!

Reply:
Thanks, Justin! That looks useful; I'll try them later.

Reply:
There doesn't seem to be any photon events, at least in the first run. This is still useful (e.g. there are some "command" messages in there), but we'll need something more real.

Reply:
Hmm, @josephhanrahan looked at all of these. There are not many events, but there should be _some._ He was able to reconstruct the scan length with his codes and show that there times beyond the SPIDR roll over. Joe, when you have a chance, maybe make a thread on Slack about this?

Reply:
I didn't check them all; there might be some events in other files.